flink-formats/flink-avro/src/main/java/org/apache/flink/formats/avro/AvroDeserializationSchema.java

```diff
@@ -0,0 +1,215 @@
+		this.reader = reader;
+		if (reader != null) {
+			this.schemaString = reader.toString();
+		}
+	}
+
+	/**
+	 * Creates {@link AvroDeserializationSchema} that produces {@link GenericRecord} using provided schema.
+	 *
+	 * @param schema schema of produced records
+	 * @return deserialized record in form of {@link GenericRecord}
+	 */
+	public static AvroDeserializationSchema<GenericRecord> forGeneric(Schema schema) {
+		return new AvroDeserializationSchema<>(GenericRecord.class, schema);
+	}
+
+	/**
+	 * Creates {@link AvroDeserializationSchema} that produces classes that were generated from avro schema.
+	 *
+	 * @param tClass class of record to be produced
+	 * @return deserialized record
+	 */
+	public static <T extends SpecificRecord> AvroDeserializationSchema<T> forSpecific(Class<T> tClass) {
+		return new AvroDeserializationSchema<>(tClass, null);
+	}
+
+	GenericDatumReader<T> getDatumReader() {
+		if (datumReader != null) {
+			return datumReader;
+		}
+
+		if (SpecificRecord.class.isAssignableFrom(recordClazz)) {
+			this.datumReader = new SpecificDatumReader<>();
+		} else if (GenericRecord.class.isAssignableFrom(recordClazz)) {
+			this.datumReader = new GenericDatumReader<>();
+		} else {
+			this.datumReader = new ReflectDatumReader<>();
+		}
+
+		return datumReader;
+	}
+
+	Schema getReaderSchema() {
+		if (reader != null) {
+			return reader;
+		}
+
+		if (SpecificRecord.class.isAssignableFrom(recordClazz)) {
+			this.reader = SpecificData.get().getSchema(recordClazz);
+		} else if (GenericRecord.class.isAssignableFrom(recordClazz)) {
+			throw new IllegalStateException(
+				"Cannot infer schema for generic record. Please pass explicit schema in the ctor.");
+		} else {
+			this.reader = ReflectData.get().getSchema(recordClazz);
+		}
+
+		return reader;
+	}
+
+	MutableByteArrayInputStream getInputStream() {
+		return inputStream;
+	}
+
+	Decoder getDecoder() {
+		return decoder;
+	}
+
+	@Override
+	public T deserialize(byte[] message) {
+		// read record
+		try {
+			inputStream.setBuffer(message);
+			Schema readerSchema = getReaderSchema();
+			GenericDatumReader<T> datumReader = getDatumReader();
+
+			datumReader.setSchema(readerSchema);
+
+			return datumReader.read(null, decoder);
+		} catch (Exception e) {
+			throw new RuntimeException("Failed to deserialize message.", e);
+		}
+	}
+
+	private void readObject(ObjectInputStream ois) throws ClassNotFoundException, IOException {
+		ois.defaultReadObject();
+		this.inputStream = new MutableByteArrayInputStream();
+		this.decoder = DecoderFactory.get().binaryDecoder(inputStream, null);
+		if (schemaString != null) {
+			this.reader = new Schema.Parser().parse(schemaString);
+		}
+	}
+
+	@Override
+	public boolean isEndOfStream(T nextElement) {
+		return false;
+	}
+
+	@Override
+	@SuppressWarnings("unchecked")
+	public TypeInformation<T> getProducedType() {
+		if (SpecificRecord.class.isAssignableFrom(recordClazz)) {
+			return new AvroTypeInfo(recordClazz, false);
```

Comment:
Agree raw types should be avoided. Unfortunately I cannot find a way to avoid it here. Your solution does not work cause the `recordClazz.asSubclass(..)` returns `<? extends SpecificRecord>` which does not conforms to `T` of the returned type.

Reply:
Oh, I see. I guess you can push the type info creation into a static method with a different generic scope, but that would not be simpler to read. Would keep it like it is then.